pkg/plugins/unpack_image.go

```diff
@@ -0,0 +1,153 @@
+package plugins
+
+import (
+	"context"
+	"fmt"
+	"github.com/containerd/containerd/archive"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/daemon"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/mudler/yip/pkg/logger"
+	"github.com/mudler/yip/pkg/schema"
+	"github.com/twpayne/go-vfs/v4"
+	"net/http"
+	"os"
+	"runtime"
+	"syscall"
+)
+
+func UnpackImage(l logger.Interface, s schema.Stage, fs vfs.FS, console Console) error {
+	if len(s.UnpackImages) == 0 {
+		return nil
+	}
+	for _, imageConf := range s.UnpackImages {
+		if imageConf.Source == "" {
+			l.Warn("No source defined for unpack_image")
+			continue
+		}
+		if imageConf.Target == "" {
+			l.Warn("No target defined for unpack_image")
+			continue
+		}
+		// create the target directory if it doesnt exist
+		if err := mkdirAll(fs, imageConf.Target, 0755); err != nil {
+			l.Errorf("Error creating target directory for unpack_image: %w", err)
+			continue
+		}
+		// unpack the image
+		image, err := getImage(imageConf.Source, imageConf.Platform)
+		if err != nil {
+			l.Errorf("Error getting image for unpack_image: %w", err)
+			continue
+		}
+		if err := extractOCIImage(image, imageConf.Target); err != nil {
+			l.Errorf("Error extracting image for unpack_image: %w", err)
+			continue
+		}
+	}
+	return nil
+}
+
+// ExtractOCIImage will extract a given targetImage into a given targetDestination
+func extractOCIImage(img v1.Image, targetDestination string) error {
+	reader := mutate.Extract(img)
+
+	_, err := archive.Apply(context.Background(), targetDestination, reader)
+
+	return err
+}
+
+func getImage(targetImage, targetPlatform string) (v1.Image, error) {
+	var platform *v1.Platform
+	var image v1.Image
+	var err error
+
+	if targetPlatform != "" {
+		platform, err = v1.ParsePlatform(targetPlatform)
+		if err != nil {
+			return image, err
+		}
+	} else {
+		platform, err = v1.ParsePlatform(fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH))
+		if err != nil {
+			return image, err
+		}
+	}
+
+	ref, err := name.ParseReference(targetImage)
+	if err != nil {
+		return image, err
+	}
+
+	tr := transport.NewRetry(http.DefaultTransport)
+	image, err = daemon.Image(ref)
+
+	if err != nil {
+		opts := []remote.Option{
+			remote.WithTransport(tr),
+			remote.WithPlatform(*platform),
+		}
+
+		opts = append(opts, remote.WithAuthFromKeychain(authn.DefaultKeychain))
```

Comment:
Still no idea how this works with auth and such. Maybe an auth method or something may be required on teh unpack conf? Will this use the containerd facilities to authenticate? no idea yet, but supporting authentication seems like a basic feature for oci artifacts.